### PR TITLE
JAMES-4085 [S3 SSEC] Guice binding & document

### PR DESCRIPTION
**Pull request by vttranlina**

### CODE REVIEW

File: docs/modules/servers/partials/configure/blobstore.adoc [hunk 1]
```diff
@@ -126,6 +126,29 @@ BucketPrefix is the prefix of bucket names in James BlobStore
...
+| optional: Boolean. Default is false. Controls whether to use Server-Side Encryption with Customer-Provided Keys (SSE-C) for S3 blobs.
+
+| encryption.s3.sse.c.master.key.algorithm
+| String. Required if `encryption.s3.sse.c.enable` is true. The algorithm used to derive the master key from the provided password. Eg: AES256
+
+| encryption.s3.sse.c.key.derivation.enabled
+| optional: Boolean. Default is false. Controls whether to derive the encryption key from the bucket name as master key and blob-id as salt.
+
+| encryption.s3.sse.c.master.key.password
+| String. Required if `encryption.s3.sse.c.key.derivation.enabled` is false. The password used to derive the customer key.
```
Comment (quantranhong1999) [inline docs/modules/servers/partials/configure/blobstore.adoc:145]: This makes me confused. 
`Required if encryption.s3.sse.c.key.derivation.enabled is false` => `encryption.s3.sse.c.master.key.password` is needed when key derivation is **disabled**.

Then reading the remaining part `The password used to derive the customer key` => why it is used to derive when derivation was disabled before?

I think it was 2 different "derivations", no? Then we should rephrase a bit the `The password used to derive the customer key.` part IMO.

File: server/apps/distributed-app/src/test/java/org/apache/james/WithS3SSECTest.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ ****************************************************************/
+
+package org.apache.james;
+
+import org.apache.james.blob.objectstorage.aws.sse.S3SSECConfiguration;
+import org.apache.james.jmap.JmapJamesServerContract;
+import org.apache.james.modules.S3SSECBlobStoreExtension;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class WithS3SSECTest implements JmapJamesServerContract, JamesServerConcreteContract {
```
Comment (quantranhong1999) [inline server/apps/distributed-app/src/test/java/org/apache/james/WithS3SSECTest.java:27]: Likely `MailsShouldBeWellReceived` is a better contract candidate as it have tests testing the mail content is well stored.

File: docs/modules/servers/partials/configure/blobstore.adoc (hunk 1 above)
Comment (quantranhong1999) [reply docs/modules/servers/partials/configure/blobstore.adoc:145]: +1

**Comment by vttranlina** [comment]
rebase & squash

**Comment by Arsnael** [comment]
Needs a rebase

File: server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3BlobStoreConfiguration.java [hunk 3]
```diff
@@ -230,7 +269,9 @@ public final boolean equals(Object o) {
                 && Objects.equals(this.writeTimeout, that.writeTimeout)
                 && Objects.equals(this.connectionTimeout, that.connectionTimeout)
                 && Objects.equals(this.uploadRetrySpec, that.uploadRetrySpec)
-                && Objects.equals(this.specificAuthConfiguration, that.specificAuthConfiguration);
+                && Objects.equals(this.specificAuthConfiguration, that.specificAuthConfiguration)
+                && Objects.equals(this.ssecEnabled, that.ssecEnabled)
+                && Objects.equals(this.ssecConfiguration, that.ssecConfiguration);
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3BlobStoreConfiguration.java:274]: I don't forget it, I still not mark resolved it, It quiet simple,
I'm wailt to pr https://github.com/apache/james-project/pull/2493 merged first, then rebase this pr,
I don't want to rebase several time

[the remark this answers is not in the document]
Comment (Arsnael) [reply server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3BlobStoreConfiguration.java:274]: Sorry I mixed up the two PRs with the comment for some reason, forget it

**Comment by vttranlina** [comment]
Rebase & resolve & fixup comment & remove code related to key derivation

File: server/container/guice/distributed/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java [hunk 4]
```diff
@@ -72,10 +80,33 @@ protected void configure() {
...
+                                  S3BlobStoreConfiguration configuration,
+                                  BlobId.Factory blobIdFactory) throws InvalidKeySpecException, NoSuchAlgorithmException {
+            if (!configuration.ssecEnabled()) {
+                return new S3BlobStoreDAO(s3ClientFactory, configuration, blobIdFactory);
+            }
+
+            S3SSECustomerKeyFactory sseCustomerKeyFactory = getS3SSECustomerKeyFactory(configuration);
+            S3RequestOption s3RequestOption = new S3RequestOption(new S3RequestOption.SSEC(true, Optional.of(sseCustomerKeyFactory)));
+            return new S3BlobStoreDAO(s3ClientFactory, configuration, blobIdFactory, s3RequestOption);
+        }
```
Comment (chibenwa) [inline server/container/guice/distributed/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java:99]: IMO we do NOT need a module chooser.

**Review by chibenwa — APPROVED** [review]

**Comment by vttranlina** [comment]
squash & rebase

**Review by quantranhong1999 — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Some guice injection issue with SMTPJamesServerTest : https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-2495/6/testReport/

**Pull request merged**